codegenerator/cli/templates/static/codegen/src/eventFetching/SourceManager.res

```diff
@@ -89,3 +94,149 @@ let fetchNext = async (
     }
   }
 }
+
+type status = Active | Stalled | Done
+
+let getSourceNewHeight = async (
+  ~source: Source.t,
+  ~currentBlockHeight,
+  ~stalledPollingInterval,
+  ~status: ref<status>,
+  ~logger,
+) => {
+  let newHeight = ref(0)
+  //Amount the retry interval is multiplied between each retry
+  let backOffMultiplicative = 2
+  let initalRetryIntervalMillis = 1000
+  //Interval after which to retry request (multiplied by backOffMultiplicative between each retry)
+  let retryIntervalMillis = ref(initalRetryIntervalMillis)
+
+  while newHeight.contents <= currentBlockHeight && status.contents !== Done {
+    try {
+      let height = await source.getHeightOrThrow()
+      newHeight := height
+      if height <= currentBlockHeight {
+        // Slowdown polling when the chain isn't progressing
+        let delayMilliseconds = if status.contents === Stalled {
+          retryIntervalMillis := stalledPollingInterval // Reset possible backOff
+          stalledPollingInterval
+        } else {
+          retryIntervalMillis := initalRetryIntervalMillis // Reset possible backOff
+          source.pollingInterval
+        }
+        await Time.resolvePromiseAfterDelay(~delayMilliseconds)
+      }
+    } catch {
+    | exn =>
+      logger->Logging.childTrace({
+        "msg": `Height retrieval from ${source.name} source failed. Retrying in ${retryIntervalMillis.contents->Int.toString}ms.`,
+        "source": source.name,
+        "error": exn->ErrorHandling.prettifyExn,
+      })
+      await Time.resolvePromiseAfterDelay(
+        ~delayMilliseconds=Pervasives.max(
+          retryIntervalMillis.contents * backOffMultiplicative,
+          60_000,
+        ),
+      )
+    }
+  }
+  newHeight.contents
+}
+
+// Polls for a block height greater than the given block number to ensure a new block is available for indexing.
+let waitForNewBlock = async (sourceManager: t, ~currentBlockHeight) => {
+  let {stalledPollingInterval, sources} = sourceManager
+
+  let logger = Logging.createChild(
+    ~params={
+      "chainId": sourceManager.activeSource.chain->ChainMap.Chain.toChainId,
+      "currentBlockHeight": currentBlockHeight,
+    },
+  )
+  logger->Logging.childTrace("Initiating check for new blocks.")
+
+  let syncSources = []
+  let fallbackSources = []
+  sources->Array.forEach(source => {
+    if (
+      source.sourceFor === Sync ||
+        // Even if the active source is a fallback, still include
+        // it to the list. So we don't wait for a timeout again
+        // if all main sync sources are still not valid
+        source === sourceManager.activeSource
```

Comment:
It might cause problems for returning from fallback to HyperSync, but it may not be a big problem when we are at the head and fallback has smaller latency.